### ISSUE TITLE
Allow non-special treatment of Prelude identifiers

### ISSUE THREAD

**Issue by m0rphism** [atom-haskell/language-haskell#85]
I get different syntax highlighting for functions from the `Prelude` (e.g. `const`) as for other functions.

This is highly confusing to me as this suggests that there is a semantic difference between these functions, which there isn't. Furthermore, it is possible to hide the default `Prelude`, which then leads to highlighting functions differently, which by accident have the same name as a `Prelude` function.

I'm very new to the atom infrastructure, but I guess this is related to `grammars/haskell.cson`, which defines a `support.function.prelude.haskell` pattern, which matches Prelude function names.

I would be happy, if it could be considered to make this behavior at least optional in the plugin configuration.

Thanks for the great work!

*edit:* I've just realized, that there is an easy workaround by adding the following to the `styles.less` file:

```css
atom-text-editor::shadow .support.function.prelude.haskell {
  color: white;
}
```

However, this needs to be updated when changing the syntax theme, and new users have to first look at the sources to find the `.support.function.prelude.haskell` selector. So I think it would still be a good idea to add switch in the plugin options.

*edit2:* there seems to be similar special treatment of `Prelude` constructors in the `support.tag.haskell` pattern.

**Comment by lierdakil** [comment]
The decision to highlight prelude functions was not mine to make initially,
but I reckon it is a correct one. Many highlighting grammars treat standard
library a little bit differently from the rest. And what's more standard
than Prelude? A small caveat with Haskell is it's impossible to
differentiate arbitrary functions from arbitrary values based on just
syntax, with exception of predefined list. Since value is technically a
zero-argument function, and functions are technically values, this
distinction isn't that important, but it leads to much more visual
disparity between arbitrary functions and Prelude ones. Still, I believe
it's better than no highlighting at all.

Code highlighting isn't necessarliy about semantic differences, as it is
about cognitive differences (since it serves to make reading and
understanding the code easier). And there is indeed a cognitive difference
between Prelude functions/types/typeclasses and, ahem, other ones. I think
at least any intermediate-level Haskell programmer knows most, if not all,
Prelude identifiers by heart. Those are "common", "uninteresting"
identifiers. Everyone's seen those a million times already. So you don't
really need to read into those too much. Being highlighted in a different
color helps with this quite a bit. As for your argument on hiding Prelude,
there are relatively few valid reasons for that IMO. Besides, if you define
a function with the same name as Prelude function, I sincerely hope it
behaves the same as Prelude function (possibly on different types), because
otherwise I would call that intentional obfuscation.

So, long story short, I'm not too keen on removing this. As for making this
optional... I don't believe Atom supports grammar configuration.

What you can do, however, is redefine style for any syntactic element,
including support.function.prelude.haskell.

Add something like this to your stylesheet (edit -> your stylesheet):

atom-text-editor::shadow .support.function.prelude.haskell {
    color: inherit;
}

**Comment by m0rphism** [comment]
Thanks for the thorough response and the proposed workaround.

However, the workaround seems to work only for the `.support.function.prelude.haskell` selector. For the other related ones (`.support.class.prelude.haskell`, `.entity.other.inherited-class.haskell`, `.support.tag.haskell`) this `inherit`s the wrong color, and for `.entity.other.inherited-class.haskell` it not only changes `Eq` constraints, but also real keywords like `deriving`.

> Many highlighting grammars treat standard
library a little bit differently from the rest. And what's more standard
than Prelude?

Correct. I would argue, that that's a mistake in other languages, too ;-)

Why is it important for me, to know whether a function comes from the `Prelude` or not? I can't remember, that I've ever had the need to distinguish that a function came specifically from the `Prelude`, rather than from any other external library I've used, e.g. `mtl`. Doing this highlighting for the `Prelude`, but not for any other external libraries, seems more confusing than helpful to me.

> Code highlighting isn't necessarliy about semantic differences, as it is
about cognitive differences (since it serves to make reading and
understanding the code easier). And there is indeed a cognitive difference
between Prelude functions/types/typeclasses and, ahem, other ones. I think
at least any intermediate-level Haskell programmer knows most, if not all,
Prelude identifiers by heart. Those are "common", "uninteresting"
identifiers. Everyone's seen those a million times already. So you don't
really need to read into those too much. Being highlighted in a different
color helps with this quite a bit.

I get how it could be useful to separate *things the user knows well* from *things the user doesn't know  well*. The problem is, that this is highly specific to the user *and* his current usecase and level of knowledge, so translating this to *Prelude* vs *Non-Prelude* is in my opinion almost never a good fit. There are beginners, who don't know the Prelude well, and most intermediate programmers and up, also consider the rest of `base`, `mtl`, `transformers`, `text`, `lens`, etc. as the same bread and butter as the `Prelude`. I imagine, that introducing a visual difference seems just arbitrary to them and most of the time draws attention to something not important: "Oh the color just changed. Is my program still correct? Have I used a keyword instead of a type instead of an expression?"

> So, long story short, I'm not too keen on removing this. As for making this
optional... I don't believe Atom supports grammar configuration.

That's unfortunate. Can you tell me how the syntax theme's recognize standard library functions? Maybe there's a better way, than making my own private fork.

Thanks again for your time and work!

**Comment by lierdakil** [comment]
> However, the workaround seems to work only for the .support.function.prelude.haskell selector. For the other related ones (.support.class.prelude.haskell, .entity.other.inherited-class.haskell, .support.tag.haskell) this inherits the wrong color, and for .entity.other.inherited-class.haskell it seems to not only change Eq constraints, but also real keywords like deriving.

That, I believe, I could fix on my end. That said, I'd prefer to avoid extra complexity that would be introduced, but I'll see what I can do with that and get back to you.

> translating this to Prelude vs Non-Prelude is in my opinion almost never a good fit. 

Well, unless there's a full Haskell parser (and parts of interpreter) built into syntax highlighter, which is never going to happen for performance reasons, some compromises have to be made. I could just as well include any library into the list (that's one line in grammar generator), but that would open a whole new can of worms (like 'why isn't my favorite library included'). Prelude definitions are imported *automatically* unless you explicitly told GHC (or whichever compiler you're using) not to. So all in all, I believe Prelude vs non-Prelude is a reasonable choice. I wish I could do that for *symbols in module scope* instead, but that's impossible right now (and unless something in Atom's grammar parser changes drastically, probably ever). In any case, you're the first one to complain about this particular design decision in 2,5 years -- and that's a fact to consider.

> Can you tell me how the syntax themes recognize standard library functions?

FYI, there's a difference between syntax *theme* and syntax *grammar*. Theme only defines some colors etc for predefined scopes. And grammar actually decides what scopes go where. So long story short, this grammar just has a predefined list of Prelude identifiers (pulled from latest GHC), and assigns those a special scope. A particular rule is unimaginatively called 'prelude', so I think you should be able to monkey-patch it after package loads in your init script.

Something like this might work:

```

**Comment by lierdakil** [comment]
`atom.packages.onDidActivatePackage` allows to hook after package activation. Can't tell from the top of my head, when is grammar actually loaded into repository, but I would guess it's on package activation

**Comment by lierdakil** [comment]
Okay, I can make `inherit` trick work for all prelude-related stuff, if I just move some definitions around. Just give me some time to carefully fix some specs that fail due to related changes, and we'll be good.

**Comment by lierdakil** [comment]
I just pushed 1.9.6 :tada:, which, amongst other things, adds proper parent scopes for prelude stuff (should've done this long ago, but oh well). Also added prelude scope to prelude-related `inherited-class`. So, now you can write something like this in your stylesheet:

```less
atom-text-editor::shadow .prelude.haskell {
    color: inherit;
}
```

And it should work pretty much as one would expect. Let me know if it works for you.

**Comment by m0rphism** [comment]
> Well, unless there's a full Haskell parser (and parts of interpreter) built into syntax highlighter, which is never going to happen for performance reasons, some compromises have to be made. I could just as well include any library into the list (that's one line in grammar generator), but that would open a whole new can of worms (like 'why isn't my favorite library included').

I completely agree. Personally, I'd just strongly prefer to make the compromise in the other direction, and treat all identifiers uniformly :-)

> In any case, you're the first one to complain about this particular design decision in 2,5 years -- and that's a fact to consider.

Interesting, I would have guessed, that the crowd would be more diverse regarding this issue. Maybe a lot of people with similar concerns are using emacs or are silent about the issue, but of course it's also possible, that I'm really a minority here. Anyways, I can understand that this is a fact to consider.

> Okay, I can make inherit trick work for all prelude-related stuff, if I just move some definitions around. Just give me some time to carefully fix some specs that fail due to related changes, and we'll be good.

Awesome! Thank you very much! I really appreciate your openness despite our different opinions :-)

> I just pushed 1.9.6 [..] And it should work pretty much as one would expect. Let me know if it works for you.

:tada: I'll let you know, thanks again! :+1:

**Comment by m0rphism** [comment]
Works as expected! :tada:

It might be useful to other users if this possibility is mentioned somewhere or made into a plugin option (now that it's not grammar related anymore, maybe this is possible?). I leave this open, in case you want this issue as an reminder; feel free to close it anytime you want; I'm very happy now :-)

**Comment by lierdakil** [comment]
I seem to have forgotten to add `prelude` to `support.tag.haskell` scope, so another patch release is imminent >_>

As for the rest, I'm updating readme right now. I don't think it's strictly possible to make a toggle option in settings for this, but it's definitely possible to make a small package that would override a stylesheet. I doubt that's necessary though, since that's literally 3 lines in user config.

**Comment by lierdakil** [comment]
See if you'd like to add something: https://github.com/atom-haskell/language-haskell#special-prelude-treatment

I'm going to let this sit here for a day or so and then push it to atom.io.

**Comment by m0rphism** [comment]
Great! I have nothing to add :+1:

**Comment by lierdakil** [comment]
I've played around a bit with Atom's internals, and here's what I came up with: https://atom.io/packages/language-haskell-scoped

Thought you might be interested in testing.

Bear in mind that it's pretty hackish and basically of pre-alpha quality.
